### ISSUE TITLE
[JENKINS-35783] REST API must only be authenticated using JWT

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/blueocean-plugin#2751]
<p>In Scope</p>
<ul>
	<li>Change authentication so that it works via JWT only</li>
	<li>Remove authentication via cookie</li>
	<li>Ensure that the frontend continues to work when JWT is enabled (You may need to open another ticket to get help from the frontend team)</li>
	<li>Solution should be cURL friendly</li>
</ul>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay">jamesdumay</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35783" target="_blank">REST API must only be authenticated using JWT</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=imeredith">imeredith</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Blocker
<li><b>component(s)</b>: blueocean-plugin
<li><b>label(s)</b>: api, blueocean-imported
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2016-08-31T09:09:11+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 2
<li><b>imported</b>: 2025-12-01
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>In Scope</p>
<ul>
	<li>Change authentication so that it works via JWT only</li>
	<li>Remove authentication via cookie</li>
	<li>Ensure that the frontend continues to work when JWT is enabled (You may need to open another ticket to get help from the frontend team)</li>
	<li>Solution should be cURL friendly</li>
</ul>
</pre>
</details>
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/bwalding.png" title="bwalding's avatar" /> bwalding:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35783?focusedId=261310&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-261310">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>My base position would be to leave Basic Auth disabled until you have a concrete use case.</p>

<p>I don't think general convenience is a valid use-case - as it re-opens the door on CSRF attacks.</p>

<p>(you might be able to get around this by doing Bearer Authentication headers rather than Basic Auth headers - but understand the use-case first)</p>
</pre>
</details>

</ul>
</details>

<p>My base position would be to leave Basic Auth disabled until you have a concrete use case.</p>

<p>I don't think general convenience is a valid use-case - as it re-opens the door on CSRF attacks.</p>

<p>(you might be able to get around this by doing Bearer Authentication headers rather than Basic Auth headers - but understand the use-case first)</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35783?focusedId=261241&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-261241">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=bwalding" class="user-hover" rel="bwalding">bwalding</a> removed HTTP auth for now.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=bwalding" class="user-hover" rel="bwalding">bwalding</a> removed HTTP auth for now.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>imeredith:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35783?focusedId=261258&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-261258">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=bwalding" class="user-hover" rel="bwalding">bwalding</a> why does basic auth open csrf on an application/json poat?</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=bwalding" class="user-hover" rel="bwalding">bwalding</a> why does basic auth open csrf on an application/json poat?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35783?focusedId=261240&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-261240">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Only thing I can think of is the URL scheme but you'd need a password for that e.g.  <a href="http://demo:securepassword@blueocean.io/blue/rest/" class="external-link" target="_blank" rel="nofollow noopener">http://demo:securepassword@blueocean.io/blue/rest/</a></p>
</pre>
</details>

</ul>
</details>

<p>Only thing I can think of is the URL scheme but you'd need a password for that e.g.  <a href="http://demo:securepassword@​blueocean.io/blue/rest/" class="external-link" target="_blank" rel="nofollow noopener">http://demo:securepassword@​blueocean.io/blue/rest/</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>imeredith:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35783?focusedId=261321&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-261321">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>well, the browser automatically sends the header in csrf. Which isnt true for JWT or Bearer. But we specifically require application/json header on the rest api to protect agaist csrf,</p>

<p>Not that I am against JWT exactly, its just not as curl friendly i think? if it is curl friendly ignore me.</p>
</pre>
</details>

</ul>
</details>

<p>well, the browser automatically sends the header in csrf. Which isnt true for JWT or Bearer. But we specifically require application/json header on the rest api to protect agaist csrf,</p>

<p>Not that I am against JWT exactly, its just not as curl friendly i think? if it is curl friendly ignore me.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/bwalding.png" title="bwalding's avatar" /> bwalding:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35783?focusedId=261329&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-261329">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Because hackers.com can make a Javascript call and the Basic Auth credentials will be sent to mycorporatesite.com if the user has entered them recently. </p>

<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>$.ajax({
    type: "POST",
    url: "https://mycorporatesite.com/blue/delete-all-my-stuff",
    data: "",
    xhrFields: {
                withCredentials: true
            },
    success: function(s) { console.log('success', s); },
    dataType: 'application/json'
  });
</pre>
</div></div>
</pre>
</details>

</ul>
</details>

<p>Because hackers.com can make a Javascript call and the Basic Auth credentials will be sent to mycorporatesite.com if the user has entered them recently. </p>



```
$.ajax({
    type: "POST",
    url: "https://mycorporatesite.com/blue/delete-all-my-stuff",
    data: "",
    xhrFields: {
withCredentials: true
    },
    success: function(s) { console.log('success', s); },
    dataType: 'application/json'
  });

```

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/bwalding.png" title="bwalding's avatar" /> bwalding:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35783?focusedId=261287&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-261287">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Requiring a header only stops the most basic attacks.  Once you have XSS in the mix it does nothing.</p>
</pre>
</details>

</ul>
</details>

<p>Requiring a header only stops the most basic attacks.  Once you have XSS in the mix it does nothing.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>imeredith:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35783?focusedId=261322&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-261322">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Ok i can go for that.</p>
</pre>
</details>

</ul>
</details>

<p>Ok i can go for that.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/bwalding.png" title="bwalding's avatar" /> bwalding:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35783?focusedId=261340&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-261340">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Some comments from a recent HN post about JWT and curl.</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>As a developer currently working with OpenID Connect (OIDC) and JSON Web Token (JWT), using curl is indeed not a problem at all:
    curl -i ... // Perform authentication to obtain JWT
    export JWT="eY..." // Place JWT in a shell variable
    curl -i -H "Authorization: Bearer $JWT" ... // Call your API
That's all there is to it.
reply
	
80x25 5 hours ago

Exactly, perhaps even simpler:
   curl ... // perform authentication to obtain JWT | xargs -I TOKEN curl -H "Authorization: Bearer TOKEN" ... // Call your API

</pre>
</div></div>

<p>Not quite as simple as a single curl, but probably simple enough.</p>
</pre>
</details>

</ul>
</details>

<p>Some comments from a recent HN post about JWT and curl.</p>


```
As a developer currently working with OpenID Connect (OIDC) and JSON Web Token (JWT), using curl is indeed not a problem at all:
    curl -i ... // Perform authentication to obtain JWT
    export JWT="eY..." // Place JWT in a shell variable
    curl -i -H "Authorization: Bearer $JWT" ... // Call your API
That's all there is to it.
reply
	
80x25 5 hours ago

Exactly, perhaps even simpler:
   curl ... // perform authentication to obtain JWT | xargs -I TOKEN curl -H "Authorization: Bearer TOKEN" ... // Call your API


```

<p>Not quite as simple as a single curl, but probably simple enough.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/vivek.png" title="vivek's avatar" /> vivek:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35783?focusedId=261288&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-261288">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Yeah lets keep basic auth aside, it can leave hole and you don't want to use it in browser client. Something on JWT vs cookie I read couple years ago: <a href="http://www.slideshare.net/derekperkins/authentication-cookies-vs-jwts-and-why-youre-doing-it-wrong" class="external-link" target="_blank" rel="nofollow noopener">http://www.slideshare.net/derekperkins/authentication-cookies-vs-jwts-and-why-youre-doing-it-wrong</a>. </p>
</pre>
</details>

</ul>
</details>

<p>Yeah lets keep basic auth aside, it can leave hole and you don't want to use it in browser client. Something on JWT vs cookie I read couple years ago: <a href="http://www.slideshare.net/derekperkins/authentication-cookies-vs-jwts-and-why-youre-doing-it-wrong" class="external-link" target="_blank" rel="nofollow noopener">http://www.slideshare.net/derekperkins/authentication-cookies-vs-jwts-and-why-youre-doing-it-wrong</a>. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/vivek.png" title="vivek's avatar" /> vivek:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35783?focusedId=266141&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-266141">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>All backend API changes are in place along with tests. More testing needs to be done though. Although its closer to ready for review for merge but we need all the UI side changes for end to end to work. Yes JWT is enabled by default and can be disabled by providing system property so its important to finish UI side changes and do a single merge.</p>

<p>See all the fronted and backend tasks at <a href="https://github.com/jenkinsci/blueocean-plugin/pull/392" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/blueocean-plugin/pull/392</a>.</p>
</pre>
</details>

</ul>
</details>

<p>All backend API changes are in place along with tests. More testing needs to be done though. Although its closer to ready for review for merge but we need all the UI side changes for end to end to work. Yes JWT is enabled by default and can be disabled by providing system property so its important to finish UI side changes and do a single merge.</p>

<p>See all the fronted and backend tasks at <a href="https://github.com/jenkinsci/blueocean-plugin/pull/392" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/blueocean-plugin/pull/392</a>.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>imeredith:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35783?focusedId=267475&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-267475">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>This is blocked on frontend tests breaking. Wi'll use meeting tomorrow to talka about it.</p>
</pre>
</details>

</ul>
</details>

<p>This is blocked on frontend tests breaking. Wi'll use meeting tomorrow to talka about it.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/tfennelly.png" title="tfennelly's avatar" /> tfennelly:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35783?focusedId=268090&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-268090">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Use of <tt>pem-jwk</tt> NPM package is causing a bit of grief for us because it's published code has a <tt>const</tt> variable declaration, which is not liked by <tt>browserify-transform-tools</tt> (which we use in js-builder). I've raised <a href="https://github.com/benbria/browserify-transform-tools/issues/23" class="external-link" target="_blank" rel="nofollow noopener">a ticket on that</a> so hope to find a solution.</p>
</pre>
</details>

</ul>
</details>

<p>Use of <tt>pem-jwk</tt> NPM package is causing a bit of grief for us because it's published code has a <tt>const</tt> variable declaration, which is not liked by <tt>browserify-transform-tools</tt> (which we use in js-builder). I've raised <a href="https://github.com/benbria/browserify-transform-tools/issues/23" class="external-link" target="_blank" rel="nofollow noopener">a ticket on that</a> so hope to find a solution.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/tfennelly.png" title="tfennelly's avatar" /> tfennelly:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35783?focusedId=268290&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-268290">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=imeredith" class="user-hover" rel="imeredith">imeredith</a> PR to fix the issues you were having: <a href="https://github.com/jenkinsci/blueocean-plugin/pull/444" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/blueocean-plugin/pull/444</a></p>

<p>I fixed the immediate issues cause by the <tt>pem-jwk</tt> NPM package using <tt>const</tt> in the published code, but I think this is going to cause issues on some browsers (def IE) and we'll have to sort that too.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=imeredith" class="user-hover" rel="imeredith">imeredith</a> PR to fix the issues you were having: <a href="https://github.com/jenkinsci/blueocean-plugin/pull/444" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/blueocean-plugin/pull/444</a></p>

<p>I fixed the immediate issues cause by the <tt>pem-jwk</tt> NPM package using <tt>const</tt> in the published code, but I think this is going to cause issues on some browsers (def IE) and we'll have to sort that too.</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `blocks` from Jira: <a href="https://github.com/jenkinsci/blueocean-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-36851%22">JENKINS-36851</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `blocks` from Jira: <a href="https://github.com/jenkinsci/blueocean-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-36852%22">JENKINS-36852</a>]</i>
